### PR TITLE
Refactor into separate subcrates, API refactor

### PR DESCRIPTION
**Pull request by dwerner**
I've done some significant rework of the repo, and most of these choices are off the cuff and suiting me in my own use case, so I don't expect them to be everyone's cup of tea. However I did the work of separating things into distinct mods, as well as added a procedural macro which I think DRYs up the addition of new commands significantly. 

So in summary:

- fixed a segfault in freeing a null ptr
- added support for extracting Reply::Error and Reply::Array types from redis
- bumped crate version to 0.2.0
- separated Command into RedisCommand and RedisCommandAttrs, allowing part to be derived
- `#[derive(RedisCommandAttrs)` support in a procedural macro (redis-command-gen subcrate)
- updated redismodule.h to support redis 4.0-rc3 (current)
- added call4 for variadic `call` with arity of 4 (though I'm not sure I use it)

In conclusion, I think that we're now 90% of the way to extracting a useful crate to publish on crates.io, but I didn't get around to that, and wanted to see how this would be received before I went ahead and did that.

I'd love to know people's thoughts.

### CODE REVIEW

**Comment by dwerner** [comment]
Obviously I need to learn some more about travis, however I was unable tonight to get the tests running for linux, because of what I suspect is a toolchain issue with the ci scripts. I think it's related to the trust stuff? Sorry for breaking this...

**Comment by brandur** [comment]
Hi @dwerner, I just wanted to drop a note in to say that this is great! I've had a busy couple of weeks and haven't been able to take a very close look yet, but will soon. Thanks for putting all this together!

**Comment by dwerner** [comment]
Thanks man, and please, if I can improve things, please let me know.

**Comment by dwerner** [comment]
I added a `redis_module!()` macro to ease creation of RedisModule_OnLoad in a consumer's crate. As an example of using the modules as they are, I wrote https://github.com/dwerner/redis-copyhash.

**Comment by KenanSulayman** [comment]
Genius! That just took a significant amount of friction out of redis module development for me. Thanks, @dwerner.

🍺

**Comment by brandur** [comment]
Hey @dwerner, apologies again for the extended wait on this one.

I'm interested in trying to get this pulled into mainline. Would you mind taking a pass over it yourself to see if there's anything in here that you might yourself change now that Rust has moved on by a few months?

Aside from that, here's a couple high level comments just to get us started off here:

1. Given that we've now branched off into more modularity, what do you think about trying to add a few module-specific tests? I think some of `redis-command-sys` is going to be pretty tough because you need an online Redis, but it might be nice to have a small sample Redis command for `redis-command-gen` module so that we exercise it's fairly considerable macro and also provide a basic usage example as a secondary benefit.
2. I think it would be nice for the new `redis-command-gen` and `redis-command-sys` crates to have a description string in their `Cargo.toml`. I think the naming uses terms that are a little too general for people to get a good grasp on them.
3. Could we get some documentation somewhere demonstrating how to use the gen crate, the sys crate, or both? Possibly just something light in individual README files.
4. Regarding the `command` API, the `static_handle="CL_THROTTLE_COMMAND"` struck me as a slight leak in implementation details. Really users probably only care about the commands they're creating, and might be confused that they have to add arbitrary secondary identifiers to them to pass to `redis_module!`. Do you think it'd be possible to `redis_module!` to take command names like `redis_module!("redis-cell", 1, "cl.throttle");` or struct names like `redis_module!("redis-cell", 1, ThrottleCommand)` (or something like that)?

Once again, thanks for very significant work on the project!

**Comment by tlvenn** [comment]
Hi @dwerner , any update on your side regarding latest comment from @brandur  ? Thanks in advance.

**Comment by gsquire** [comment]
Chiming in a bit late but since the goal is to simplify the bindings to the C headers I wanted to point out that we could leverage the [bindgen](https://github.com/rust-lang-nursery/rust-bindgen) crate. The only annoyance is that we will still need to write a small wrapper around the static init function that has already been done here.

**Comment by dpc** [comment]
@dwerner  @brandur 

It seems a bit stale now. I wonder if there's any chance of any action. I think I'm going to take the cell + store bits and fork them into separate crate because that's the only part I care about it, and it can be reused in projects that are not even `redis` related.

Any comments?

**Comment by dwerner** [comment]
I think the partial extraction of redis specific stuff should live in it's own crate, I just never got around to it. If you make progress, it would be cool to contribute to that.

**Comment by dpc** [comment]
@dwerner https://github.com/dpc/ratelimit-gcra

**Comment by dpc** [comment]
@dwerner
I looked at this PR, and it looks like a right approach. However, as I only care about algo. itself, and redis part is unnecessary for me, I am not planing to do anything here other than extracting the algo itself.

 I am planning to publish the algo crate on crates.io, as I need it for my own project. If anyone is interested in owning it, I am happy to hand it over.
